File: maximumsubarray.go
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
	we know current contiguous max sum if we know continuous max sum of previous index
	DP[i] = DP[i - 1] + nums[i] if DP[i - 1] >= 0
	DP[i] = nums[i] if DP[i - 1] < 0
	ans is max between ans, curr
*/
package main

import "fmt"

func maxSubArray(nums []int) int {
	ans := nums[0]
	curr := nums[0]
	for i := 1; i < len(nums); i++ {
		if curr >= 0 {
			curr += nums[i]
		} else {
			curr = nums[i]
		}
		ans = max(ans, curr)
	}
	return ans
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func MainMaximumSubArray() {
	fmt.Println(maxSubArray([]int{-2, 1, -3, 4, -1, 2, 1, -5, 4})) //expected 6
}

File: Maximum-subarray.cpp
/*Time Complexity: O(N)

Space Complexity: O(1)
Did this code successfully run on Leetcode : Yes

Approach: Kadens algo - consider the subarray's with positive sum only.

Prob: 53. Maximum Subarray
*/
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int n = nums.size();
        long long maxi = LONG_MIN;
        long long sum = 0;
        for(int i=0;i<n;i++){
            sum += nums[i];
            if(sum > maxi){
                maxi = max(sum,maxi);
            }
            if(sum < 0){
                sum = 0;
            }
        }
        return maxi;
    }
};